### DISCUSSION TITLE
need help with variables and secrets

### DISCUSSION THREAD

**Discussion by sureshdevops99** [community/community#52497] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

I would like to know about below questions.

How can we use runner or container variables within github actions script? Like how do we use them inside the github actions plugins?
what is the difference between env defined in the github actions script and env (list of variables) of runner/container?
How can we create a temporary secrets or runtime secrets within the scope of github actions script and not storing in the org/repo level secrets?

**Comment by foysal619** [comment]
Using runner or container variables within GitHub Actions script:
To use runner or container variables within a GitHub Actions script, you can use the syntax ${{ env.VARIABLE_NAME }} to access the variable. For example, if you have a runner or container variable named MY_VARIABLE, you can use it in a step like this:

yaml

- name: My Step
  run: |
    echo ${{ env.MY_VARIABLE }}
This will output the value of MY_VARIABLE within the runner or container.

If you want to use runner or container variables in a GitHub Actions plugin, you can use the same syntax within the plugin's configuration. For example, if you're using the checkout plugin and you want to use a runner or container variable as part of the path where the repository is checked out, you can do it like this:

yaml

- name: Checkout code
  uses: actions/checkout@v2
  with:
    ref: ${{ github.ref }}
    path: ${{ env.MY_VARIABLE }}/my-project
Difference between env defined in the GitHub Actions script and env (list of variables) of runner/container:
The env defined in the GitHub Actions script is a way to define environment variables that are specific to the workflow, and they are available to all steps within the workflow. On the other hand, the env of the runner/container is a list of environment variables that are specific to the runner or container where the workflow is running. These variables are available to all workflows running on that runner or container.

The env of the runner/container is usually defined by the system administrator or by the user who sets up the runner, and it can be used to set environment variables that are required by the runner or container itself. For example, you might use this to set the JAVA_HOME environment variable for a runner that needs to run a Java application.

Creating temporary secrets or runtime secrets within the scope of GitHub Actions script:
GitHub Actions allows you to create temporary secrets that are only available during the execution of a workflow. To create a temporary secret, you can use the set-env command to set an environment variable with the value of the secret.

For example, suppose you want to create a temporary secret named MY_SECRET with the value my-secret-value. You can do it like this:

yaml

- name: Set temporary secret
  run: |
    echo "::set-env name=MY_SECRET::my-secret-value"
Once the workflow finishes, the temporary secret will no longer be available.

Note that this approach should only be used for secrets that do not need to persist beyond the execution of the workflow. If you need to store a secret for future use, you should use the GitHub Actions secrets feature to store the secret securely.

**Reply by airtower-luna** [reply]
This answer smells of ChatGPT. There is no such thing as a "temporary secret" in the Actions syntax. The `set-env` command *used to be* a way to set environment variables until it was replaced with an environment file due to security issues. (There are probably more issues, but I'm not going to bother after the obvious ones.)

**Comment by airtower-luna** [comment]
> How can we use runner or container variables within github actions script? Like how do we use them inside the github actions plugins?

What exactly do you mean by "runner or container variables"? Also, what do you mean by "github actions script" (maybe the workflow syntax?) and "github actions plugins" (maybe custom actions)?

> what is the difference between env defined in the github actions script and env (list of variables) of runner/container?

You can define environment variables at different levels of a workflow: For the whole workflow, a job, a step, or at runtime during a job. The variable will be available for the context for which it is defined.

> How can we create a temporary secrets or runtime secrets within the scope of github actions script and not storing in the org/repo level secrets?

You can [mask secret values](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#masking-a-value-in-a-log), which will hide them in the log like secrets. Mind the warning in the linked documentation:

> Warning: Make sure you register the secret with 'add-mask' before outputting it in the build logs or using it in any other workflow commands.

**Comment by sureshdevops99** [comment]
Hi [airtower-luna] & [foysal619]

Basically, I am aware of variables we will create in the workflow/script, 
for ex: 
name: JavaWebApp
on: 
  workflow_dispatch: 
env:
  foo: bar
  abc: 123

or

name: JavaWebApp
on: 
  workflow_dispatch: 
jobs:
  build:
     runs-on: ubuntu-latest
     env:
      MY_NAME: "Test Name"

I have an environment variable inside the runner/container used for running the workflow
for ex: foo=bar or abc=123
now, how can I use this in the github actions workflow/script

**Comment by airtower-luna** [comment]
Okay, so you're looking at environment variables. :slightly_smiling_face: 

> I have an environment variable inside the runner/container used for running the workflow
> for ex: foo=bar or abc=123

Where exactly do you have those environment variables? Written in a file? Defined in the shell of a job step? Or do you have a self-hosted runner where they are defined elsewhere and want to transfer the variables into your workflow? Or something else?

**Comment by sureshdevops99** [comment]
I have these variables in the self hosted runner environment variables. 
env
foo=bar

I want to use them in the actions - for ex:
    - name: Test Variable
      env:
        var1: $foo
      run: echo $foo

output:
bar

**Comment by airtower-luna** [comment]
Okay, that's the trickiest option. I don't see anything about that in the runner documentation, and (without knowing the design docs) I'd expect the runner to *avoid* depending on the environment it runs in. I'd say if it's an option in your context, avoid that approach, and define the variables elsewhere (e.g. repository secrets or variables, or a file in the repository).

If you absolutely need to transfer the variables, I think your best option is to write them  to a file in the [format used for the environment file](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable) (e.g. using a cron job, but that depends on how the variables are defined on your machine), and append the content of that file to the file `$GITHUB_ENV` points at early in your job.
